### PR TITLE
[Backport] [v1.2.5] Cache the engines and volumes object to speed up the upgrade path from v1.2.2 to v1.2.3

### PR DESCRIPTION
**Pull request by PhanLe1010**
longhorn/longhorn#3980
longhorn/longhorn#3996

### CODE REVIEW

File: upgrade/v122to123/upgrade.go [hunk 1]
```diff
@@ -42,6 +42,24 @@ func upgradeBackups(namespace string, lhClient *lhclientset.Clientset) (err erro
 		return err
 	}
 
+	engines, err := lhClient.LonghornV1beta1().Engines(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	volumeNameToEngines := make(map[string][]longhorn.Engine)
```
Comment (shuo-wu) [inline upgrade/v122to123/upgrade.go:49]: NIT: is it better to use pointers here? `map[string][]*longhorn.Engine`. Similar below

**Review by shuo-wu — APPROVED** [review]
LGTM

File: upgrade/v122to123/upgrade.go (hunk 1 above)
Comment (PhanLe1010) [reply upgrade/v122to123/upgrade.go]: Fixed it. I has been always scare of using pointer :D

**Review by innobead — APPROVED** [review]

File: upgrade/v122to123/upgrade.go (hunk 1 above)
Comment (shuo-wu) [reply upgrade/v122to123/upgrade.go]: It's better not to the pointers of objects init by the `for` loop. IIRC, the for loop here can be considered as:
```
var e longhorn.Engine
for idx := 0; idx < len(engines.Items); i++ {
    e = engines.Items[idx]
    p := &e  // This always points to the same variable, which may be problematic
}
```

For more detail, you can try this: https://go.dev/play/p/tjB7R-znss6

Comment (shuo-wu) [reply upgrade/v122to123/upgrade.go]: This kind of syntactic sugar is tricky anyway.

Comment (PhanLe1010) [reply upgrade/v122to123/upgrade.go]: Screw me! My big mistake.
 Many thanks for the alarm!

Comment (PhanLe1010) [reply upgrade/v122to123/upgrade.go]: @shuo-wu 
PR created at https://github.com/longhorn/longhorn-manager/pull/1340

**Pull request merged**